spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala

```diff
@@ -0,0 +1,390 @@
+            rays.add(message)
+          }
+        }
+      })
+    }
+
+    val info: Seq[PointInfo] = {
+      val fn = pointInfo(elevation)_
+      ps.zipWithIndex.map(fn)
+    }
+
+    val _pointsByKey: Map[SpatialKey, Seq[PointInfo]] =
+      info
+        .groupBy(_.key)
+        .toMap
+    val pointsByKey = sc.broadcast(_pointsByKey)
+    if (touchedKeys != null) touchedKeys ++= _pointsByKey.keys
+
+    val _pointsByIndex: Map[Int, PointInfo] =
+      info
+        .groupBy(_.index)
+        .mapValues({ list => list.head })
+        .toMap
+    val pointsByIndex = sc.broadcast(_pointsByIndex)
+
+    val _heightsByIndex: Map[Int, Double] = // index -> height
+      elevation
+        .flatMap({ case (k, v) =>
+          val key = implicitly[SpatialKey](k)
+          val tile = implicitly[Tile](v)
+
+          pointsByKey.value.get(key) match {
+            case Some(list) =>
+              list.map({ case PointInfo(index, _, col, row, viewHeight0, _, _, _) =>
+                val viewHeight =
+                  if (viewHeight0 >= 0.0) tile.getDouble(col, row) + viewHeight0 ; else -viewHeight0
+                (index, viewHeight)
+              })
+            case None => Seq.empty[(Int, Double)]
+          }
+        })
+        .collect
+        .toMap
+    val heightsByIndex = sc.broadcast(_heightsByIndex)
+
+    // Create RDD  of viewsheds; after this,  the accumulator contains
+    // the rays emanating from the starting points.
+    var sheds: RDD[(K, V, MutableArrayTile)] = elevation.map({ case (k, v) =>
+      val key = implicitly[SpatialKey](k)
+      val tile = implicitly[Tile](v)
+      val shed = R2Viewshed.generateEmptyViewshedTile(tile.cols, tile.rows)
+
+      pointsByKey.value.get(key) match {
+        case Some(list) =>
+          list.foreach({ case PointInfo(index, _, col, row, _, ang, fov, alt) =>
+            val viewHeight = heightsByIndex.value.getOrElse(index, throw new Exception)
+
+            R2Viewshed.compute(
+              tile, shed,
+              col, row, viewHeight,
+              FromInside(),
+              null,
+              rayCatcherFn(key, index),
+              resolution = resolution,
+              maxDistance = maxDistance,
+              curvature = curvature,
+              altitude = alt,
+              operator = operator,
+              cameraDirection = ang,
+              cameraFOV = fov
+            )
+          })
+        case None =>
+      }
+
+      (k, v, shed)
+    }).persist(StorageLevel.MEMORY_AND_DISK_SER)
+    sheds.count // make sheds materialize
+
+    // Repeatedly  map over the RDD  of viewshed tiles until  all rays
+    // have reached the periphery of the layer.
+    do {
+      val _changes: Map[SpatialKey, Seq[(Int, From, mutable.ArrayBuffer[Ray])]] =
+        rays.value
+          .groupBy(_._1)
+          .map({ case (k, list) =>
+            (k, list.map({ case (_, index, from, rs) => (index, from, rs) }))
+          })
+          .toMap
```

Comment:
How big are these change sets typically? It seems like this is this sorting, and the sort by index few lines below, can happen in the accumulator easily enough but not clear if the optimization is likely to be material.